juniper_warp/src/lib.rs

```diff
@@ -289,68 +317,243 @@ where
+                            payload.variables,
+                        );
+
+                        let mut executor = juniper::SubscriptionsExecutor::<'_, Context, S>::new();
+                        let response_stream = graphql_request
+                            .subscribe_async(&schema, &context, &mut executor)
+                            .await;
+
+                        if let Some(error) = response_stream.errors() {
+                            println!("Error occured: {:#?}", error);
+                            panic!("Response stream is none (error occured)");
+                        }
+
+                        let stream = response_stream
+                            .into_stream()
+                            .expect("Response stream is none");
+
+                        stream
+                            .take_while(move |response| {
+                                let request_id = request_id.clone();
+                                let closed = got_close_signal.load(Ordering::Relaxed);
+                                if closed {
+                                    let close_text = format!(
+                                        r#"{{"type":"complete","id":"{}","payload":null}}"#,
+                                        request_id
+                                    );
+
+                                    // send message that we are closing channel
+                                    let _ = ws_tx.unbounded_send(Some(Ok(Message::text(
+                                        close_text.clone(),
+                                    ))));
+
+                                    // close channel
+                                    let _ = ws_tx.unbounded_send(None);
+                                } else {
+                                    let mut response_text =
+                                        serde_json::to_string(&response).unwrap();
+                                    response_text = format!(
+                                        r#"{{"type":"data","id":"{}","payload":{} }}"#,
+                                        request_id, response_text
+                                    );
+
+                                    let _ = ws_tx.unbounded_send(Some(Ok(Message::text(
+                                        response_text.clone(),
+                                    ))));
+                                }
+                                async move { !closed }
+                            })
+                            .for_each(|_| async {})
+                            .await;
+                    });
+                }
+                "stop" => {
+                    got_close_signal.store(true, Ordering::Relaxed);
+                }
+                _ => panic!("unknown type"),
+            }
+
+            futures03::future::ready(())
+        })
+        .for_each(|_| {
+            // empty `for_each` here to keep executing this stream
+            async {}
```

Comment:
Why is this needed? This seems pretty "magical", I wonder if there are helpers in the `futures` crate.